File: Assignment7_3/src/main/java/application/controller/CurrencyConverterController.java
package application.controller;

import application.model.CurrencyCollector;
import application.view.CurrencyConverterView;
import dao.CurrencyDao;
import javafx.application.Platform;
import javafx.fxml.FXMLLoader;
import javafx.scene.Scene;
import javafx.stage.Stage;

import java.io.IOException;
import java.sql.SQLException;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

public class CurrencyConverterController {

    private CurrencyCollector currencyCollector;
    private CurrencyConverterView view;

    public CurrencyConverterController(CurrencyConverterView view) {
        this.view = view;
        currencyCollector = new CurrencyCollector();

        ArrayList<String> tempList = new ArrayList<>(getCurrencies());
        tempList.addFirst("Select");
        ArrayList<String> tempList1 = new ArrayList<>(tempList);
        view.setChoiceBoxSource(tempList);
        view.setChoiceBoxTarget(tempList1);
        view.setSourcePlaceholder("Select Currency");
        view.setTargetPlaceholder("Select Currency");
        view.setChoiceBoxSourceValue("Select");
        view.setChoiceBoxTargetValue("Select");
    }

    public List<String> getCurrencies() {
        try {
            return currencyCollector.getCurrencies();
        } catch (SQLException e) {
            view.showSQLError();
        }
        return new ArrayList<>();
    }

    /**
     * convert value from one currency to another
     * @param c1 source currency abbreviation
     * @param c2 target currency abbreviation
     * @param value value
     */
    public void convert(String c1, String c2, Double value) {
        try {
            double result = currencyCollector.convertTo(c1, c2, value);
            if (Double.isInfinite(result)) {
                throw new NumberFormatException("Cannot divide by zero");
            } else if (Double.isNaN(result)) {
                throw new NumberFormatException("Value is not a number");
            }
            DecimalFormat df = new DecimalFormat("0.00####");
            view.setTarget(df.format(result));
        } catch (SQLException e) {
            view.showSQLError();
        } catch (NumberFormatException e) {
            view.showAlert();
        }
    }

    public void convertButtonPressed() {
        try {
            String source = view.getSource();
            if (source.matches(".*[,].*") && !source.matches(".*[.].*")) {
                source = source.replaceAll("[,]", ".");
            }
            Double value = Double.parseDouble(source);
            view.setSource(String.format("%.2f", value));
            String[] choiceBoxes = view.getChoiceBoxes();
            this.convert(choiceBoxes[0], choiceBoxes[1], value);
        } catch (Exception e) {
            view.showAlert();
        }
    }

    public void onSourceChoiceBoxChange() {
        try {
            view.setSourcePlaceholder(currencyCollector.getCurrencyName(view.getChoiceBoxes()[0]));
        } catch (SQLException e) {
            view.showSQLError();
        }
    }

    public void onTargetChoiceBoxChange() {
        try {
            view.setTargetPlaceholder(currencyCollector.getCurrencyName(view.getChoiceBoxes()[1]));
        } catch (SQLException e) {
            view.showSQLError();
        }
    }

    public void newCurrency() throws IOException {
        Stage stage = new Stage();
        FXMLLoader loader = new FXMLLoader(getClass().getResource("/view/newCurrency_view.fxml"));
        Scene scene = new Scene(loader.load());
        NewCurrencyViewController controller = loader.getController();
        controller.init(stage, currencyCollector.getDao(), view);
        stage.setScene(scene);
        stage.showAndWait();
        Platform.runLater(() -> {
            view.setChoiceBoxTarget(new ArrayList<>(getCurrencies()));
            view.setChoiceBoxSource(new ArrayList<>(getCurrencies()));
            view.setChoiceBoxTargetValue(getCurrencies().get(0));
            view.setChoiceBoxSourceValue(getCurrencies().get(0));
        });
    }
}

File: Assignment7_2_Kotlin/src/main/kotlin/application/controller/CurrencyConverterController.kt
package application.controller

import application.model.CurrencyCollector
import application.view.CurrencyConverterView
import java.sql.SQLException
import java.text.DecimalFormat

class CurrencyConverterController(private val view: CurrencyConverterView) {
    private val currencyCollector: CurrencyCollector

    init {
        currencyCollector = CurrencyCollector()

        val tempList = ArrayList<String?>(this.currencies)
        tempList.addFirst("Select")
        val tempList1 = ArrayList<String?>(tempList)
        view.setChoiceBoxSource(tempList)
        view.setChoiceBoxTarget(tempList1)
        view.setSourcePlaceholder("Select Currency")
        view.setTargetPlaceholder("Select Currency")
        view.setChoiceBoxSourceValue("Select")
        view.setChoiceBoxTargetValue("Select")
    }

    val currencies: MutableList<String?>?
        get() {
            try {
                return currencyCollector.currencies
            } catch (e: SQLException) {
                view.showSQLError()
            }
            return ArrayList<String?>()
        }

    /**
     * convert value from one currency to another
     * @param c1 source currency abbreviation
     * @param c2 target currency abbreviation
     * @param value value
     */
    fun convert(c1: String?, c2: String?, value: Double) {
        try {
            val result: Double = currencyCollector.convertTo(c1, c2, value)
            println(result)
            val df = DecimalFormat("0.00####")
            view.setTarget(df.format(result))
        } catch (e: SQLException) {
            view.showSQLError()
        }
    }

    fun convertButtonPressed() {
        try {
            var source: String = view.getSource()!!
            if (source.matches(".*[,].*".toRegex()) && !source.matches(".*[.].*".toRegex())) {
                source = source.replace("[,]".toRegex(), ".")
            }
            val value = source.toDouble()
            view.setSource(String.format("%.2f", value))
            val choiceBoxes: Array<String?> = view.choiceBoxes
            this.convert(choiceBoxes[0], choiceBoxes[1], value)
        } catch (e: Exception) {
            view.showAlert()
        }
    }

    fun onSourceChoiceBoxChange() {
        try {
            view.setSourcePlaceholder(currencyCollector.getCurrencyName(view.choiceBoxes[0]))
        } catch (e: SQLException) {
            view.showSQLError()
        }
    }

    fun onTargetChoiceBoxChange() {
        try {
            view.setTargetPlaceholder(currencyCollector.getCurrencyName(view.choiceBoxes[1]))
        } catch (e: SQLException) {
            view.showSQLError()
        }
    }
}